Tests/iaas/image-metadata/image-md-check.py

```diff
@@ -121,17 +145,70 @@ def is_ok(self, props, warn = ""):
 
 
 def is_url(stg):
-    "Is string stg a URL?"
+    """Is string stg a URL?"""
     idx = stg.find("://")
-    if idx < 0:
-        return False
-    if stg[:idx] in ("http", "https", "ftp", "ftps"):
-        return True
-    return False
+    return idx >= 0 and stg[:idx] in ("http", "https", "ftp", "ftps")
+
+
+def parse_date(stg, formats=DATE_FORMATS):
+    """
+    Return time in Unix seconds or 0 if stg is not a valid date.
+    We recognize: %Y-%m-%dT%H:%M:%SZ, %Y-%m-%d %H:%M[:%S], and %Y-%m-%d
+    """
+    bdate = 0
+    for fmt in formats:
+        try:
+            tmdate = time.strptime(stg, fmt)
+            bdate = calendar.timegm(tmdate)
+            break
+        except ValueError:  # as exc:
+            # print(f'date {stg} does not match {fmt}\n{exc}', file=sys.stderr)
+            pass
+    return bdate
+
+
+def freq2secs(stg):
+    """Convert frequency to seconds (round up a bit), return 0 if not applicable"""
+    secs = FREQ_TO_SEC.get(stg)
+    if secs is None:
+        print(f'ERROR: replace frequency {stg}?', file=sys.stderr)
+        secs = 0
+    return secs
+
+
+OUTDATED_IMAGES = []
+
+
+def is_outdated(img, bdate):
+    """return 1 if img (with build/regdate bdate) is outdated,
+       2 if it's not hidden or marked, 3 if error"""
+    max_age = 0
+    if "replace_frequency" in img.properties:
+        max_age = 1.1 * (freq2secs(img.properties["replace_frequency"]))
+    if not max_age or time.time() <= max_age + bdate:
+        return 0
+    # So we found an outdated image that should have been updated
+    # (5a1) Check whether we are past the provided_until date
+    until_str = img.properties["provided_until"]
+    until = parse_date(img.properties["provided_until"])
+    if not until and not until_str == "none" and not until_str == "notice":
+        print(f'ERROR: Image "{img.name}" does not provide a valid provided until date',
+              file=sys.stderr)
+        return 3
+    if time.time() > until:
+        return 0
+    if img.is_hidden or img.name[-3:] == "old" or img.name[-4:] == "prev" or img.name[-8:].isdecimal():
+        return 1
```

Comment:
The standard states

> It is recommended that the old image gets renamed (e.g. build date or patch level attached)

There's no mention of "old" nor "prev". On the other hand, a patch level is mentioned, and that will not be recognized by `is_outdated`, will it?

I think we need to work this out more, but I'm afraid this would need a more strict standard: either make the `os_hidden` mandatory, or mandate some other clear marker (such as a new property `original_name` that has to be set to the name of the current version, which must be a prefix of the name of the image in question).

Reply:
A funny observation: if the patch level is "Gold", then it will be recognized ;)

Reply:
You are right, the standard should define better defined rules for renaming (or hiding) old images, so users can rely on it and the test can have a less heuristic approach to checking it.
Patch level does not have a defined format either, so we can't use it unless we define a format.
I am afraid this will require a v2 of the image metadata standard.

Reply:
Moving forward short-term without a v2 standard yet (and thus in absence of strict rules that make the test reliably avoid false positives), I see a few things to decide.
* We could drop the "prev" "old" "yyyymmdd" heuristic piece -- this will cause all outdated images to end up on a list where newer images are searched. (Or use this heuristic piece only to trigger a warning or not, but not decide whether the image ends up on the list.) Actually I think we should do this!
* When outdated images are found without a replacement being recognized by the script, we currently emit an error. We could downgrade this to a warning to avoid false positives. On the other hand, all existing SCS clouds pass this and we could probably live with the risk and wait for cloud providers to really trigger this ... and then react by either improving the algorithm further or then transforming the error into a warning. (I prefer this.)

Reply:
Sorry, I forgot to say that I agree.